File: evaluation/S0/javascript/Medium/1059-string-without-aaa-or-bbb.js
// Source: https://leetcode.com/problems/string-without-aaa-or-bbb/   |   Difficulty: Medium
//
// Problem Description:
// Given two integers a and b, return any string s such that:
//
//
// 	s has length a + b and contains exactly a 'a' letters, and exactly b 'b' letters,
// 	The substring 'aaa' does not occur in s, and
// 	The substring 'bbb' does not occur in s.
//
// Example:
// Input: a = 1, b = 2
// Output: "abb"
// Explanation: "abb", "bab" and "bba" are all correct answers.
//
// Constraints:
// 0 <= a, b <= 100
// 	It is guaranteed such an s exists for the given a and b.
//

/**
 * @param {number} a
 * @param {number} b
 * @return {string}
 */
var strWithout3a3b = function(a, b) {
    
    let result = [];
    while (a > 0 || b > 0) {
        if (a > b) {    
            if (a > 1) {
                result.push('aa');
                a -= 2;
            } else {
                result.push('a');
                a -= 1;
            }   
            if (b > 0) {
                result.push('b');
                b -= 1;
            }

        } else if (b > a) {
            if (b > 1) {
                result.push('bb');
                b -= 2;
            }
            else {  
                result.push('b');
                b -= 1;
            }
            if (a > 0) {    


                result.push('a');                a -= 1;
            }
        } else {    
            if (a > 0) {
                result.push('a');
                a -= 1;
            }
            if (b > 0) {
                result.push('b');
                b -= 1;
            }

        }
    }
    return result.join(''); 
    
};

File: evaluation/S1/python/Medium/1059-string-without-aaa-or-bbb.py
# Source: https://leetcode.com/problems/string-without-aaa-or-bbb/   |   Difficulty: Medium
#
# Problem Description:
# Given two integers a and b, return any string s such that:
#
#
# 	s has length a + b and contains exactly a 'a' letters, and exactly b 'b' letters,
# 	The substring 'aaa' does not occur in s, and
# 	The substring 'bbb' does not occur in s.
#
# Example:
# Input: a = 1, b = 2
# Output: "abb"
# Explanation: "abb", "bab" and "bba" are all correct answers.
#
# Constraints:
# 0 <= a, b <= 100
# 	It is guaranteed such an s exists for the given a and b.
#
# Helpful references (internal KB):
# - Suffix Automaton (graph, string, simulation)
#   • When to use: Use when you need to efficiently represent all distinct substrings of a string, or perform various string operations like counting unique substrings, finding occurrences, or string matching in an online fashion.
#   • Idea: A suffix automaton is a powerful data structure that represents all distinct substrings of a given string in a compact form, typically as a Directed Acyclic Word Graph (DAWG). It can be constructed in linear time O(N) and space O(N) for a string of length N.
#   • Invariants: Each state in the automaton represents a set of substrings that share the same end-positions in the original string.; For any state `v`, `len(v)` is the length of the longest string represented by `v`.
#   • Tips: Understand the properties of `len` (length of the longest string in a state) and `link` (suffix link to the longest proper suffix).; The construction is online, processing characters one by one from left to right.
#   • Pitfalls: Incorrectly handling the `link` updates, especially when cloning states during construction.; Forgetting to update the `last` pointer after each character addition step.
# - Linear Diophantine Equation (number, greedy)
#   • When to use: Use when a problem can be solved by making the locally optimal choice at each step, hoping this leads to a globally optimal solution. This strategy is often simpler and faster than dynamic programming.
#   • Idea: The greedy algorithm makes the best choice at each stage with the hope of finding a global optimum, without reconsidering previous choices. Its time complexity is typically O(N log N) or O(N).
#   • Invariants: At each step, the chosen element is the best available according to the greedy criterion.; The partial solution constructed so far remains optimal for the subproblem considered.
#   • Tips: Prove correctness by exchange argument or induction.; Sort input data if it helps identify the greedy choice.
#   • Pitfalls: Failing to prove the greedy choice property.; Making a locally optimal choice that prevents a global optimum.
# - Generating all K-combinations (array, enumeration, greedy)
#   • When to use: Use this when you need to iterate through all possible subsets of a fixed size K from a larger set, especially when a specific order like lexicographical is required.
#   • Idea: This algorithm systematically generates all unique combinations of K elements from a set of N elements. It typically proceeds in lexicographical order, allowing for efficient enumeration of all C(N, K) possibilities.
#   • Invariants: The current combination C is always sorted: C[0] < C[1] < ... < C[K-1].; All elements C[i] are distinct and within the range [0, N-1] (assuming 0-indexed).
#   • Tips: Start with the smallest lexicographical combination (e.g., {1, 2, ..., K}).; To find the next combination, locate the rightmost element that can be incremented.
#   • Pitfalls: Incorrectly handling boundary conditions when incrementing elements.; Duplicating combinations if not careful with the increment logic.
#
# Notes:
# - Keep the given function signature unchanged.
# - Use only the Python standard library; no third-party packages.
# - Do not change input/output formats or return value semantics as defined by the template/problem.
# - Conformance: use features compatible with LeetCode’s Python3 runtime; do NOT rely on undefined/implementation-specific behavior.
# - If you introduce any helper functions/classes, include their full implementations in this file; no hidden helpers.

class Solution(object):
    def strWithout3a3b(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: str
        """
        result = []
        while a > 0 or b > 0:
            if a > b:
                if a >= 2:
                    result.append('aa')
                    a -= 2
                else:
                    result.append('a')
                    a -= 1
                if b > 0:
                    result.append('b')
                    b -= 1
            elif b > a:
                if b >= 2:
                    result.append('bb')
                    b -= 2
                else:
                    result.append('b')
                    b -= 1
                if a > 0:
                    result.append('a')
                    a -= 1
            else:  # a == b
                if a > 0:
                    result.append('ab')
                    a -= 1
                    b -= 1
        return ''.join(result)